File: Java/src/main/java/edu/rice/dmodel/Supplier.java
package edu.rice.dmodel;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

import javax.json.Json;
import javax.json.JsonObject;
import javax.json.JsonObjectBuilder;
import javax.json.JsonReader;
import javax.json.JsonWriter;
import javax.json.JsonWriterFactory;
import javax.json.stream.JsonGenerator;

import com.fasterxml.jackson.core.JsonToken;
import com.google.protobuf.CodedInputStream;

import edu.rice.pdb.serialization.BsonSinglton;
//import edu.rice.pdb.serialization.ProtocolBufferSinglton;
import edu.rice.pdb.serialization.proto.CustomerProtos.CustomerP.OrderP.LineItemP.SupplierP;

public class Supplier extends Base implements RootData  {

	/**
	 * This class represents a supplier as it is specified in TPC-H benchmark
	 * schema
	 */
	private static final long serialVersionUID = -2215947411317295533L;

	// CREATE TABLE [dbo].[SUPPLIER](
	// [S_SUPPKEY] [int] NOT NULL,
	// [S_NAME] [char](25) NOT NULL,
	// [S_ADDRESS] [varchar](40) NOT NULL,
	// [S_NATIONKEY] [int] NOT NULL,
	// [S_PHONE] [char](15) NOT NULL,
	// [S_ACCTBAL] [decimal](15, 2) NOT NULL,
	// [S_COMMENT] [varchar](101) NOT NULL
	// );

	private int supplierKey;
	private String name;
	private String address;
	private int nationKey;
	private String phone;
	private double accbal;
	private String comment;

	public Supplier() {

	}

	public Supplier(int supplierKey, String name, String address, int nationKey, String phone, double accbal, String comment) {
		super();
		this.supplierKey = supplierKey;
		this.name = name;
		this.address = address;
		this.nationKey = nationKey;
		this.phone = phone;
		this.accbal = accbal;
		this.comment = comment;
	}
	
	public Supplier generateSupplier() {
		int i = 0;
		return	new Supplier( i, "name" + i, "address" + i, i, "phone" + i, i, "comment"+ i);
		
	}

	public ArrayList<RootData> generateObjects(int number) {

		ArrayList<RootData> objectList = new ArrayList<RootData>();

		for (int i = 0; i < number; i++) {
			Supplier tmp = new Supplier( i, "name" + i, "address" + i, i, "phone" + i, i, "comment"+ i);
			objectList.add(tmp);
		}

		return objectList;
	}

	public JsonObjectBuilder jsonJsonObjectBuilder() {
		JsonObjectBuilder supplierBuilder = Json.createObjectBuilder();

		supplierBuilder.add("name", this.getName())
		               .add("supplierkey", this.getSupplierKey())
		               .add("address", this.getAddress())
		               .add("nationkey", this.getNationKey())
		               .add("phone", this.getPhone())
		               .add("accbal", this.getAccbal())
		               .add("comment", this.getComment());

		return supplierBuilder;
	}

	public Supplier readJSONSupplier(JsonObject supplierJsonObject) {

		Supplier m_supplier = new Supplier();
		m_supplier.setName(supplierJsonObject.getString("name"));
		m_supplier.setSupplierKey(supplierJsonObject.getInt("supplierkey"));
		m_supplier.setAddress(supplierJsonObject.getString("address"));
		m_supplier.setAccbal(supplierJsonObject.getInt("accbal"));
		m_supplier.setNationKey(supplierJsonObject.getInt("nationkey"));
		m_supplier.setComment(supplierJsonObject.getString("comment"));
		m_supplier.setPhone(supplierJsonObject.getString("phone"));

		return m_supplier;
	}

	public int getSupplierKey() {
		return supplierKey;
	}

	public void setSupplierKey(int supplierKey) {
		this.supplierKey = supplierKey;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getAddress() {
		return address;
	}

	public void setAddress(String address) {
		this.address = address;
	}

	public int getNationKey() {
		return nationKey;
	}

	public void setNationKey(int nationKey) {
		this.nationKey = nationKey;
	}

	public String getPhone() {
		return phone;
	}

	public void setPhone(String phone) {
		this.phone = phone;
	}

	public double getAccbal() {
		return accbal;
	}

	public void setAccbal(double accbal) {
		this.accbal = accbal;
	}

	public String getComment() {
		return comment;
	}

	public void setComment(String comment) {
		this.comment = comment;
	}

	public String toString() {

		String s = this.supplierKey + " -" + this.name + " -" + this.address + " -" + this.nationKey + " -" + this.phone + " -"
				+ this.accbal + " -" + this.comment;
		return s;
	}
	
	
	 public byte[] writeByteBuffer(){

//			private int supplierKey; // 4 bytes 
//			private int nationKey;  // 4 bytes 
//			private double accbal;  // 8 bytes 
 
		 
//			private String name; 
//			private String address;
//			private String phone;
//			private String comment;

		 
		    byte[] nameBytes = name.getBytes(Charset.forName("UTF-8"));
			byte[] addressBytes = address.getBytes(Charset.forName("UTF-8"));
			byte[] phoneBytes = phone.getBytes(Charset.forName("UTF-8"));
			byte[] commentBytes = comment.getBytes(Charset.forName("UTF-8"));

			// allocate a buffer to write this object
			ByteBuffer byteBuffer = ByteBuffer.allocate(32 + nameBytes.length + addressBytes.length + phoneBytes.length + commentBytes.length);

			byteBuffer.putInt(supplierKey); // 4 bytes
			byteBuffer.putInt(nationKey); // 4 bytes
			byteBuffer.putDouble(accbal); // 8 bytes


			// name
			byteBuffer.putInt(nameBytes.length); // 4 bytes size of "container string"
			byteBuffer.put(nameBytes);

			// address
			byteBuffer.putInt(addressBytes.length); // 4 bytes size of "container string"
			byteBuffer.put(addressBytes);

			// phone
			byteBuffer.putInt(phoneBytes.length); // 4 bytes size of "container string"
			byteBuffer.put(phoneBytes);
		
			// comment
			byteBuffer.putInt(commentBytes.length); // 4 bytes size of "container string"
			byteBuffer.put(commentBytes);

			return byteBuffer.array();
	 }
	 

	 public Supplier readByteBuffer(byte[] buf){
		 
		 

//			private int supplierKey; // 4 bytes 
//			private int nationKey;  // 4 bytes 
//			private double accbal;  // 8 bytes 

		 
//			private String name; 
//			private String address;
//			private String phone;
//			private String comment;
		 
		    // allocate a buffer for 12 bytes to write an integer and a double
			ByteBuffer byteBuffer = ByteBuffer.wrap(buf);

			int tmpSupplierKey = byteBuffer.getInt(); // 4 bytes
			int tmpNationKey = byteBuffer.getInt(); // 4 bytes = 8
			Double tmpAccbal = byteBuffer.getDouble(); // 8 bytes = 16

			int stringSize;

			// Get name
			stringSize = byteBuffer.getInt(); // 4 bytes = 20
			String tmpName = extractString(byteBuffer, stringSize);

			// address
			stringSize = byteBuffer.getInt(); // 4 bytes = 24
			String tmpAddress = extractString(byteBuffer, stringSize);

			// phone
			stringSize = byteBuffer.getInt(); // 4 bytes = 24
			String tmpPhone = extractString(byteBuffer, stringSize);

			// comment
			stringSize = byteBuffer.getInt(); // 4 bytes = 24
			String tmpComment = extractString(byteBuffer, stringSize);
		 return new Supplier(tmpSupplierKey, tmpName,tmpAddress, tmpNationKey, tmpPhone, tmpAccbal, tmpComment) ;
	 }

	@Override
	public int compareTo(RootData o) {
		Supplier mySupplier=(Supplier) o;
		return Integer.compare(mySupplier.getSupplierKey(), this.getSupplierKey());	
	}


	@Override
	public byte[] jsonSerialization() {
		JsonObjectBuilder supplierBuilder = this.jsonJsonObjectBuilder();;

		JsonObject partJsonObject = supplierBuilder.build();

		Map<String, Object> properties = new HashMap<String, Object>(1);
		properties.put(JsonGenerator.PRETTY_PRINTING, true);
		JsonWriterFactory writerFactory = Json.createWriterFactory(properties);

		ByteArrayOutputStream bos = new ByteArrayOutputStream();
        try {
			GZIPOutputStream gzipOutputStream = new GZIPOutputStream(bos);
			JsonWriter jsonWriter = writerFactory.createWriter(gzipOutputStream);
			jsonWriter.writeObject(partJsonObject);
			jsonWriter.close();
		} catch (IOException e) {
			logger.error(e);		
		}



		// Get the bytes of the serialized JSON object
		byte[] buf = bos.toByteArray();
		return buf;
	}

	@Override
	public RootData jsonDeserialization(byte[] buf) {
		Supplier mySupplier = new Supplier();
		ByteArrayInputStream b = new ByteArrayInputStream(buf);
		try {
			InputStream gzis = new GZIPInputStream(b);
			
			// create JsonReader object
			JsonReader jsonReader = Json.createReader(gzis);

			// get JsonObject from JsonReader
			JsonObject jsonObject = jsonReader.readObject();

			mySupplier.setName(jsonObject.getString("name"));
			mySupplier.setSupplierKey(jsonObject.getInt("supplierkey"));
			mySupplier.setAddress(jsonObject.getString("address"));
			mySupplier.setAccbal(jsonObject.getInt("accbal"));
			mySupplier.setNationKey(jsonObject.getInt("nationkey"));
			mySupplier.setComment(jsonObject.getString("comment"));
			mySupplier.setPhone(jsonObject.getString("phone"));
			
			// we can close IO resource and JsonReader now
			jsonReader.close();
		} catch (Exception e) {
			logger.error("Can not read JSON object Supplier from byteArray", e);
		}
		return mySupplier;
	}

	@Override
	public byte[] protocolBufferWrite() {
		SupplierP.Builder mySupplier = SupplierP.newBuilder();
//		SupplierP.Builder mySupplier = ProtocolBufferSinglton.getInstance().getSupplier();

		mySupplier.setSupplierKey(this.supplierKey);
		mySupplier.setName(this.name);
		mySupplier.setAddress(this.address);
		mySupplier.setAccbal(this.accbal);
		mySupplier.setNationKey(this.nationKey );
		mySupplier.setComment(this.comment);
		mySupplier.setPhone(this.phone);

		byte[] buf = null;

		ByteArrayOutputStream bos = new ByteArrayOutputStream();

		// Write the class room to disk.
		try {
			mySupplier.build().writeTo(bos);
			// Get the bytes of the serialized JSON object
			buf = bos.toByteArray();
			bos.close();

		} catch (IOException e) {
			logger.error("Protocol Buffer Writer in Supplier ", e);
		}

		return buf;
	}

	@Override
	public RootData protocolBufferRead(byte[] buf) {
		Supplier mySupplier = new Supplier();
		ByteArrayInputStream b = new ByteArrayInputStream(buf);

		SupplierP.Builder protocSupplier = SupplierP.newBuilder();
		
//		SupplierP.Builder protocSupplier = ProtocolBufferSinglton.getInstance().getSupplier();

		CodedInputStream stream = CodedInputStream.newInstance(b);

		// mergeFrom
		try {
			protocSupplier.mergeFrom(stream);
		} catch (IOException e) {
			logger.error(" Protocol Buffer Reader  in Supplier ", e);
		}

		mySupplier.setName(protocSupplier.getName());
		mySupplier.setSupplierKey(protocSupplier.getSupplierKey());
		mySupplier.setAddress(protocSupplier.getAddress());
		mySupplier.setAccbal(protocSupplier.getAccbal());
		mySupplier.setNationKey(protocSupplier.getNationKey());
		mySupplier.setComment(protocSupplier.getComment());
		mySupplier.setPhone(protocSupplier.getPhone());

		return mySupplier;
	}

	@Override
	public RootData javaDefaultDeserialization(byte[] buf) {
		ByteArrayInputStream b = new ByteArrayInputStream(buf);
		ObjectInputStream objectInputStream;

		Supplier p = null;

		try {
			objectInputStream = new ObjectInputStream(b);
			p = (Supplier) objectInputStream.readObject();

		} catch (IOException | ClassNotFoundException e) {
			logger.error("Can read object Supplier from byteArray", e);
		}
		return p;
	}

	
	/**
	 * 
	 * @return
	 */
	@Override
	public byte[] bsonSerialization() {

		// serialize data
		ByteArrayOutputStream baos = new ByteArrayOutputStream();
		try {
			com.fasterxml.jackson.core.JsonGenerator gen = BsonSinglton.getInstance().getFactory().createJsonGenerator(baos);

			gen.writeStartObject();

			gen.writeFieldName("name");
			gen.writeNumber(this.getName());

			gen.writeFieldName("supplierkey");
			gen.writeNumber(this.getSupplierKey());

			gen.writeFieldName("address");
			gen.writeString(this.getAddress());

			gen.writeFieldName("accbal");
			gen.writeNumber(this.getAccbal());

			gen.writeFieldName("nationkey");
			gen.writeNumber(this.getNationKey());

			gen.writeFieldName("phone");
			gen.writeString(this.getPhone());

			gen.writeFieldName("comment");
			gen.writeString(this.getComment());

			gen.close();

		} catch (IOException e) {
			logger.error("Part bsonSerialization ", e);
		}
		return baos.toByteArray();
	}

	/**
	 * 
	 * @param buf
	 * @param myclass
	 * @return
	 */
	@Override
	public RootData bsonDeSerialization(byte[] buf, Class<?> myclass) {

		ByteArrayInputStream bais = new ByteArrayInputStream(buf);

		Supplier tmp = new Supplier();

		try {
			com.fasterxml.jackson.core.JsonParser parser = BsonSinglton.getInstance().getFactory().createJsonParser(bais);

			parser.nextToken();

			while (parser.nextToken() != JsonToken.END_OBJECT) {
				String fieldname = parser.getCurrentName();
				parser.nextToken();

				switch (fieldname) {
				case "name":
					tmp.setName(parser.getText());
					break;
				case "supplierkey":
					tmp.setSupplierKey(parser.getIntValue());
					break;
				case "address":
					tmp.setAddress(parser.getText());
					break;
				case "accbal":
					tmp.setAccbal(parser.getDoubleValue());
					break;
				case "nationkey":
					tmp.setNationKey(parser.getIntValue());
					break;
				case "phone":
					tmp.setPhone(parser.getText());
					break;
				case "comment":
					tmp.setComment(parser.getText());
					break;
				}

			}

		} catch (IOException e) {
			logger.error("Element bsonSerialization ", e);

		}

		return tmp;
	}
	
	

}

File: C++/src/handcoded/source/Supplier.cpp
#include <iostream>
#include <cstring>
#include <sstream>
#include <vector>
#include "Supplier.h"
using namespace std;


	// CREATE TABLE [dbo].[SUPPLIER](
	// [S_SUPPKEY] [int] NOT NULL,
	// [S_NAME] [char](25) NOT NULL,
	// [S_ADDRESS] [varchar](40) NOT NULL,
	// [S_NATIONKEY] [int] NOT NULL,
	// [S_PHONE] [char](15) NOT NULL,
	// [S_ACCTBAL] [decimal](15, 2) NOT NULL,
	// [S_COMMENT] [varchar](101) NOT NULL
	// );



	//Default constructor:
	Supplier::Supplier() {		
	}


	//Constructor with arguments:
	Supplier::Supplier(int supplierKey, string name, string address, int nationKey, string phone, double accbal, string comment) {
		this->supplierKey = supplierKey;
		this->name = name;
		this->address = address;
		this->nationKey = nationKey;
		this->phone = phone;
		this->accbal = accbal;
		this->comment = comment;		
	}


	//Getters and Setters Follow:
	int Supplier::getSupplierKey() {
		return supplierKey;
	}

	void Supplier::setSupplierKey (int supplierKey) {
		this->supplierKey = supplierKey;
	}

	string Supplier::getName() {
		return name;
	}

	void Supplier::setName(string name) {
		this->name = name;
	}

	string Supplier::getAddress() {
		return address;
	}

	void Supplier::setAddress(string address) {
		this->address = address;
	}

	int Supplier::getNationKey() {
		return nationKey;
	}

	void Supplier::setNationKey(int nationKey) {
		this->nationKey = nationKey;
	}

	string Supplier::getPhone() {
		return phone;
	}

	void Supplier::setPhone(string phone) {
		this->phone = phone;
	}

	double Supplier::getAccbal() {
		return accbal;
	}

	void Supplier::setAccbal(double accbal) {
		this->accbal = accbal;
	}

	string Supplier::getComment() {
		return comment;
	}

	void Supplier::setComment(string comment) {
		this->comment = comment;
	}



	//C++: Explicit call needed for printing content:
	string Supplier::toString() {
		string stringS =  itos(this->getSupplierKey()) + " - " + this->getName() + " - " + this->getAddress() + " - " + itos(this->getNationKey()) + " - " +  this->getPhone() + " - " + dtos(this->getAccbal()) + " - " + this->getComment();	  
		return stringS;
	}	



	//Use for generating objects in bulk:
	vector<Supplier *> Supplier::generateObjects(int number) {
		vector<Supplier *> objectList;
		for (int i = 0; i < number; i++) {
			Supplier * tmp = new Supplier(i, "Name-" + itos(i), "Address-" + itos(i+1), i, "Phone-" + itos(i+2), 0.1 * i, "Comment-" + itos(i+3));
			objectList.push_back(tmp);
		}
		return objectList;
	}



	/*
	 * Functions related to Hand Coded Serialization.
	 */

	//Hand Coded C++ serialization:
	char * Supplier::serialize(char * buffer, int& objectSize) {

		//First determine the size of object:
		//Integer elements:
		objectSize = 2 * sizeof(int);
		//Double elements:
		objectSize += 1 * sizeof(double);
		//String sizes: 
		objectSize += 4 * sizeof(int);
		//String Elements:
		objectSize += (name.length() + address.length() + phone.length() + comment.length());	


		//Now, serialize the object.
		char * tempBuffer = buffer;
		
		//Copy Integers:
		tempBuffer = copyInt(tempBuffer, supplierKey);	
		tempBuffer = copyInt(tempBuffer, nationKey);	
		//Copy Doubles:		
		tempBuffer = copyDouble(tempBuffer, accbal);		
		//Copy Strings:
		tempBuffer = copyString(tempBuffer, name);		
		tempBuffer = copyString(tempBuffer, address);
		tempBuffer = copyString(tempBuffer, phone);
		tempBuffer = copyString(tempBuffer, comment);

		return buffer;
	}



	int Supplier::deserialize(char * buffer) {
		int bytesRead = 0;
		
		//supplierKey:
		supplierKey = parseInt(buffer+bytesRead);
		bytesRead += sizeof(supplierKey);	
		
		//nationKey:
		nationKey = parseInt(buffer+bytesRead);
		bytesRead += sizeof(nationKey);		
		
		//Account Balance:
		accbal = parseDouble(buffer+bytesRead);
		bytesRead += sizeof(accbal);
			
		//Parse Strings:
		parseString(buffer+bytesRead, name); 		bytesRead += (sizeof(int) + name.length());		
		parseString(buffer+bytesRead, address);		bytesRead += (sizeof(int) + address.length());		
		parseString(buffer+bytesRead, phone);		bytesRead += (sizeof(int) + phone.length());		
		parseString(buffer+bytesRead, comment);		bytesRead += (sizeof(int) + comment.length());

		return  bytesRead;
	}	


	//Implement your own custom comparator:
	bool Supplier::operator< (Supplier& other) {	
		return (supplierKey < other.supplierKey);
	}


	//Create a inplace object:
	SupplierIP * Supplier::createInPlaceObject() {
		return new SupplierIP(supplierKey, name, address, nationKey, phone, accbal, comment);		
	}

	//In place serialization:
	long Supplier::serializeInPlace(char * fileStart) {
		//In place object:
		this->inplace = this->createInPlaceObject();

		//Return pointer:
		return ((long) inplace - (long)fileStart);
	}